src/RBTree.mo

```diff
@@ -8,80 +39,286 @@ import O "Order";
+    /// does not have a return value.
+    ///
+    /// Example:
+    /// ```motoko include=initialize
+    /// import Iter "mo:base/Iter";
+    ///
+    /// tree.put(1, "one");
+    /// tree.put(2, "two");
+    ///
+    /// tree.delete(1);
+    /// Iter.toArray(tree.entries()) // now only contains [(2, "two")].
+    /// ```
+    ///
+    /// Runtime: `O(log(n))`.
+    /// Space: `O(1)` retained memory plus garbage, see the note below.
+    /// where `n` denotes the number of key-value entries stored in the tree and
+    /// assuming that the `compareTo` function implements an `O(1)` comparison.
+    ///
+    /// Note: Creates `O(log(n))` temporary objects that will be collected as garbage.
+    public func delete(key : K) {
+      let (res, t) = removeRec(key, compareTo, tree);
       tree := t
     };
 
-    /// Remove the entry associated with a given key.
-    public func remove(x : X) : ?Y {
-      let (res, t) = removeRec(x, compareTo, tree);
+    /// Remove the entry associated with a given key, if the key exists, and return the associated value.
+    /// Returns `null` without any other effect if the key is absent.
+    ///
+    /// Example:
+    /// ```motoko include=initialize
+    /// import Iter "mo:base/Iter";
+    ///
+    /// tree.put(1, "one");
+    /// tree.put(2, "two");
+    ///
+    /// let removedEntry = tree.remove(1); // => ?"one"
+    /// // tree now only contains [(2, "two")].
+    /// (removedEntry, Iter.toArray(tree.entries())) // => (?"one", [(2, "two")])
+    /// ```
+    ///
+    /// Runtime: `O(log(n))`.
+    /// Space: `O(1)` retained memory plus garbage, see the note below.
+    /// where `n` denotes the number of key-value entries stored in the tree and
+    /// assuming that the `compareTo` function implements an `O(1)` comparison.
+    ///
+    /// Note: Creates `O(log(n))` temporary objects that will be collected as garbage.
+    public func remove(key : K) : ?V {
+      let (res, t) = removeRec(key, compareTo, tree);
       tree := t;
       res
     };
 
     /// An iterator for the key-value entries of the map, in ascending key order.
+    /// The iterator takes a snapshot view of the tree and is not affected by concurrent modifications.
+    ///
+    /// Example:
+    /// ```motoko include=initialize
+    /// import Debug "mo:base/Debug";
+    ///
+    /// tree.put(1, "one");
+    /// tree.put(2, "two");
+    /// tree.put(3, "two");
+    ///
+    /// for (entry in tree.entries()) {
+    ///   Debug.print("Entry key=" # debug_show(entry.0) # " value=\"" # entry.1 #"\"");
+    /// }
+    ///
+    /// // Entry key=1 value="one"
+    /// // Entry key=2 value="two"
+    /// // Entry key=3 value="three"
+    /// ```
+    ///
+    /// Cost of iteration over all elements:
+    /// Runtime: `O(n*log(n))`.
+    /// Space: `O(1)` retained memory plus garbage, see the note below.
+    /// where `n` denotes the number of key-value entries stored in the tree.
```

Comment:
I think that this should be O(n) runtime, O(n) space.
My reasoning is that, the iterator is basically performing DFS over the tree, using the list `trees` as an explicit stack.

Am I missing something?

Reply:
Thanks a lot. This was indeed wrong. I corrected it.